src/stash.rs

```diff
@@ -0,0 +1,204 @@
+use {raw, panic, Oid, StashApplyProgress};
+use std::ffi::{CStr};
+use util::{Binding};
+use libc::{c_int, c_char, size_t, c_void};
+use build::{CheckoutBuilder};
+use std::mem;
+
+/// Stash application progress notification function.
+///
+/// Return `true` to continue processing, or `false` to
+/// abort the stash application.
+pub type StashApplyProgressCb<'a> = FnMut(StashApplyProgress) -> bool + 'a;
+
+/// This is a callback function you can provide to iterate over all the
+/// stashed states that will be invoked per entry.
+pub type StashCb<'a> = FnMut(usize, &str, &Oid) -> bool + 'a;
+
+#[allow(unused)]
+/// Stash application options structure
+pub struct StashApplyOptions<'cb> {
+    progress: Option<Box<StashApplyProgressCb<'cb>>>,
+    checkout_options: Option<CheckoutBuilder<'cb>>,
+    raw_opts: raw::git_stash_apply_options
+}
+
+impl<'cb> StashApplyOptions<'cb> {
+    /// Creates a default set of merge options.
+    pub fn new() -> StashApplyOptions<'cb> {
+        let mut opts = StashApplyOptions {
+            progress: None,
+            checkout_options: None,
+            raw_opts: unsafe { mem::zeroed() },
+        };
+        assert_eq!(unsafe {
+            raw::git_stash_apply_init_options(&mut opts.raw_opts, 1)
+        }, 0);
+        opts
+    }
+
+    /// Set stash application flag to GIT_STASH_APPLY_REINSTATE_INDEX
+    pub fn reinstantiate_index(&mut self) -> &mut StashApplyOptions<'cb> {
+        self.raw_opts.flags = raw::GIT_STASH_APPLY_REINSTATE_INDEX;
+        self
+    }
+
+    /// Options to use when writing files to the working directory
+    pub fn checkout_options(&mut self, mut opts: CheckoutBuilder<'cb>) -> &mut StashApplyOptions<'cb> {
+        unsafe {
+            opts.configure(&mut self.raw_opts.checkout_options);
+        }
```

Comment:
I think we'll want to defer this until `raw` is called because otherwise movement of the structs may mess up some internal pointers and such.